andi/andi.py

```diff
@@ -1,62 +1,302 @@
 # -*- coding: utf-8 -*-
+from collections import OrderedDict
 from typing import (
     Dict, List, Optional, Type, Callable, Union, Container,
-    get_type_hints,
-)
+    get_type_hints, Tuple, cast, MutableMapping)
 
-from andi.typeutils import get_union_args, is_union, get_globalns
+from andi.typeutils import (
+    get_union_args,
+    is_union,
+    get_globalns,
+    get_unannotated_params,
+)
+from andi.errors import CyclicDependencyError, NonProvidableError
 
 
 def inspect(func: Callable) -> Dict[str, List[Optional[Type]]]:
     """
     For each argument of the ``func`` return a list of possible types.
+    Non annotated arguments are also returned with an empty list of possible
+    types.
     """
     globalns = get_globalns(func)
     annotations = get_type_hints(func, globalns)
+    for name in get_unannotated_params(func, annotations):
+        annotations[name] = None
     annotations.pop('return', None)
-    annotations.pop('cls', None)  # FIXME: pop first argument of methods
+    annotations.pop('self', None)  # FIXME: pop first argument of methods
+    annotations.pop('cls', None)
     res = {}
     for key, tp in annotations.items():
         if is_union(tp):
             res[key] = get_union_args(tp)
         else:
-            res[key] = [tp]
+            res[key] = [] if tp is None else [tp]
```

Comment:
a tiny suggestion: the code can be a bit more readable if None case if handled in the same way as is_union case, to make it really clear that there are 3 main cases

Reply:
Makes sense. Changed here: https://github.com/scrapinghub/andi/pull/8/commits/2963eb08b7c6543312ffe67ecacfbb810c406079